File: DSA/to_revise/76 - minWindow.py
# https://leetcode.com/problems/minimum-window-substring/

from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""
        
        hm = Counter(t)
        required = len(hm)
        l, r = 0, 0
        formed = 0
        window_counts = {}
        min_len = float('inf')
        min_window = (0, 0)
        
        while r < len(s):
            character = s[r]
            window_counts[character] = window_counts.get(character, 0) + 1
            
            if character in hm and window_counts[character] == hm[character]:
                formed += 1
            
            while l <= r and formed == required:
                character = s[l]
                
                if r - l + 1 < min_len:
                    min_len = r - l + 1
                    min_window = (l, r)
                
                window_counts[character] -= 1
                # Check if the char removed is in hm and the num of character of t
                # is still satisfied (some char might have extra occurence, so
                # deducting one from it will satisfy the constrain)
                if character in hm and window_counts[character] < hm[character]:
                    formed -= 1
                
                l += 1
            
            r += 1
        
        return "" if min_len == float('inf') else s[min_window[0]:min_window[1] + 1]

File: DSA/done/76 - minWindow.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <climits>

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";

        unordered_map<char, int> hm;
        for (char c : t) hm[c]++;

        int required = hm.size();
        int l = 0, r = 0, formed = 0;
        unordered_map<char, int> window_counts;
        int min_len = INT_MAX;
        int start_idx = 0;

        while (r < s.length()) {
            char character = s[r];
            window_counts[character]++;
            if (hm.count(character) && window_counts[character] == hm[character]) formed++;

            while (l <= r && formed == required) {
                character = s[l];
                if (r - l + 1 < min_len) {
                    min_len = r - l + 1;
                    start_idx = l;
                }

                window_counts[character]--;
                if (hm.count(character) && window_counts[character] < hm[character]) formed--;
                l++;
            }
            r++;
        }

        return min_len == INT_MAX ? "" : s.substr(start_idx, min_len);
    }
};